### DISCUSSION TITLE
Google Summer of Code 2022 Challenge

### DISCUSSION THREAD

**Discussion by alabulei1** [WasmEdge/WasmEdge#1371] [General]
### Introduction

For those who want to apply for GSoC #1227  
This is a pretest for the application.

Please fork this repo and complete the challenge in your repo. Once you are done, post your repo’s GitHub link, as well as a screenshot demonstrating the customized host functions, and the rust SDK for creating your test cases, as a comment under this discussion.

### Requirements

1. Fork this repo and try to add brand new host functions. You should define and implement your host functions. FYR, I have created a template in https://github.com/WasmEdge/WasmEdge/tree/challenge/proposal. You can find the details in the last four commits.
2. After you finish the first step, you will need to create a wasm file to call your host functions. If you have no ideas about how to make this, please fork and refer to this example: https://github.com/second-state/wasmedge_hostfunctionexample_interface.
3. Then, once you implement both 1. and 2. You can use `wasmedge foo.wasm` to run your application.
4. Post your repo's GitHub link including the forked WasmEdge and the wasm file or the rust interface, a screenshot of the execution result, and your comments under this discussion.

Good luck!

**Comment by moksh-pathak** [comment]
I have begun working on this, would be giving the updates soon.

**Comment by 2019zhou** [comment]
Hi,
I have defined and implemented my host functions called fibonacci with two functions "fibonacci_get_num", "fibonacci_get_series" and hope these are enough for the challenge.

screenshot:
![PBMeuQXiDI](https://user-images.githubusercontent.com/52394242/161591036-fbf26275-82d6-4d03-b862-1e5b882653d1.png)

my repo link: https://github.com/2019zhou/WasmEdge

Thanks for the clear and detailed instructions. If there is anything else, please let me know. 😆

**Comment by abhinandanudupa** [comment]
Hello!
As specified, I have added a custom module `InsertionSortExampleModule` which has functions to initialise an vector with randomn integers, sort them using Insertion Sort and to print the vector elements.

**Here is a screen shot of the output:**
![2022-04-18_21-21](https://user-images.githubusercontent.com/67597710/163835112-ce103a13-12d9-4910-be7f-63ed49779485.png)

**Links to the repos:**
[WasmEdge](https://github.com/abhinandanudupa/WasmEdge/tree/challenge/proposal)
[Rust Library](https://github.com/abhinandanudupa/wasmedge_hostfunctionexample_interface) 

Can you please tell me what you would expect in the proposal? My understanding so far is that we have to build a interface to the windows OS using the native functions - probably [these](https://docs.microsoft.com/en-us/windows/win32/apiindex/windows-api-list) (eg those for IO, File processing, process management etc. and similar to that which has been completed for [Linux and mac OS](https://github.com/WasmEdge/WasmEdge/blob/master/lib/host/wasmedge_process/processfunc.cpp#L177)). Hope I am right on this.

**Reply by hydai** [reply]
This challenge is going to test the understanding of using WasmEdge host functions, including defining an unique interface in WasmEdge side and writing a demo with the corresponding Rust library. WASI is one of our built-in host function sets. The goal of our GSoC mission is to use Windows API in these host functions to make them available on the Windows platform.

**Reply by abhinandanudupa** [reply]
Got it
